akka-stream-tests/src/test/scala/akka/stream/impl/fusing/GraphInterpreterSpec.scala

```diff
@@ -0,0 +1,551 @@
+      val sink2 = new DownstreamProbe[Int]("sink2")
+      val balance = new Balance[Int](2)
+
+      val assembly = GraphAssembly(
+        Array(balance.in, sink1.in, sink2.in),
+        Array(source.out, balance.out(0), balance.out(1)))
+
+      val interpreter = new GraphInterpreter(assembly)
+      interpreter.init()
+
+      lastEvents() should ===(Set.empty)
+
+      sink1.requestOne()
+      lastEvents() should ===(Set(RequestOne(source)))
+
+      sink2.requestOne()
+      lastEvents() should ===(Set.empty)
+
+      source.onNext(1)
+      lastEvents() should ===(Set(OnNext(sink1, 1), RequestOne(source)))
+
+      source.onNext(2)
+      lastEvents() should ===(Set(OnNext(sink2, 2)))
+    }
+
+    "implement merge-balance" in pending
+
+    "implement balance-merge" in pending
+
+    "implement bidi-stage" in pending
+
+    "implement non-divergent cycle" in new TestSetup {
+      val source = new UpstreamProbe[Int]("source")
+      val sink = new DownstreamProbe[Int]("sink")
+      val merge = new Merge[Int](2)
+      val balance = new Balance[Int](2)
+
+      val assembly = GraphAssembly(
+        Array(merge.in(0), merge.in(1), balance.in, sink.in),
+        Array(source.out, balance.out(1), merge.out, balance.out(0)))
+
+      val interpreter = new GraphInterpreter(assembly)
+      interpreter.init()
+
+      lastEvents() should ===(Set.empty)
+
+      sink.requestOne()
+      lastEvents() should ===(Set(RequestOne(source)))
+
+      source.onNext(1)
+      lastEvents() should ===(Set(OnNext(sink, 1), RequestOne(source)))
+
+      // Token enters merge-balance cycle and gets stuck
+      source.onNext(2)
+      lastEvents() should ===(Set(RequestOne(source)))
+
+      // Unstuck it
+      sink.requestOne()
+      lastEvents() should ===(Set(OnNext(sink, 2)))
+
+    }
+
+    "implement divergent cycle" in new TestSetup {
+      val source = new UpstreamProbe[Int]("source")
+      val sink = new DownstreamProbe[Int]("sink")
+      val merge = new Merge[Int](2)
+      val balance = new Balance[Int](2)
+      val detacher = new Detacher[Int]
+
+      val assembly = GraphAssembly(
+        Array(merge.in(0), merge.in(1), detacher.in, balance.in, sink.in),
+        Array(source.out, detacher.out, balance.out(1), merge.out, balance.out(0)))
+
+      val interpreter = new GraphInterpreter(assembly)
+      interpreter.init()
+
+      lastEvents() should ===(Set.empty)
+
+      sink.requestOne()
+      lastEvents() should ===(Set(RequestOne(source)))
+
+      source.onNext(1)
+      lastEvents() should ===(Set(OnNext(sink, 1), RequestOne(source)))
+
+      // Token enters merge-balance cycle and spins until event limit
+      source.onNext(2)
+      lastEvents() should ===(Set(RequestOne(source)))
+
+      sink.requestOne()
+      lastEvents() should ===(Set(OnNext(sink, 2)))
+
+      source.onNext(3)
+
+      //      // Unstuck it
+      //      sink.requestOne()
+      //      lastEvents() should ===(Set(OnNext(sink, 2)))
```

Comment:
this is the cycle that I intended to be divergent but it is not. This branch is still WIP so things are in an intermediate state.